### ISSUE TITLE
Nodes / Edges Disappearing 

### ISSUE THREAD

**Issue by harish-madugula** [wbkd/react-flow#494]
Hey,

I don't know if I'm doing something wrong but when my flow has more than 10 or 15 Nodes, sometimes a Node gets disappeared,
when I try to find out what's wrong, for some reason, its `opacity` was set to 0 - to solve the issue, I had to explicitly set the `opacity` to 1 for the node using the classes

```javascript
.react-flow__renderer {
        .react-flow__nodes {
            .react-flow__node {
                opacity: 1 !important;
          }
       }
}
```
And sometimes, an Edge gets disappeared.

I dont see the edge connecting the first two nodes inside the elements.

After a lot o struggle and a lot of inspecting elements, 
I found out that, inside that <g></g> tag, there are number of <g> tags, where each one is an edge.

pic 1 
![flow1](https://user-images.githubusercontent.com/28500085/92276834-623e0100-ef0f-11ea-93cf-093b430aadc3.JPG)
[pic 1](https://user-images.githubusercontent.com/28500085/92276834-623e0100-ef0f-11ea-93cf-093b430aadc3.JPG)
PIC 1
point 1 is the **edge disappearance**
point 2 is the **edge between Node 2 and Node 3**
point 3 is the styles that I alter - uncheck padding or font size

but when I alter some styling - like unchecking the padding or font-size styles, the edge connecting the first two nodes gets added from nowhere. and also visible on the UI.
PIC 2 
point 6  altering of styles (unchecked padding)
point 4 **edge appears**
point 5 - a new <g> tag appears gets added to the list of elements.

if you see the below pic,

![flow2](https://user-images.githubusercontent.com/28500085/92277138-f4dea000-ef0f-11ea-9fd4-17529ac97e8c.JPG)
[pic 2](https://user-images.githubusercontent.com/28500085/92277138-f4dea000-ef0f-11ea-9fd4-17529ac97e8c.JPG)

did anyone face this issue? or anywhere similar to this?
any help or pointers are very much appreciated.

I really wish I could get a way to reproduce this issue for you guys.
AGAIN - THIS HAPPENS ONLY WHEN THERE ARE MORE NO.OF NODES - LIKE MORE THAN 10 OR 15

Thanks in advance

**Comment by moklick** [comment]
How can we reproduce this. In this example https://reactflow.dev/stress we have 100 nodes and everything works fine.

**Comment by 100le** [comment]
I had the following issue once i stored the positions and re-rendered them again later on.
I noticed that **some of the Nodes** are missing and the edges are pointing to some strange positions in the graph but others were okey.
The problem was because i was using the .toFixed(2) which converted the Number in String and later when i tried to re-render the graph using the string position values some of the Nodes were gone.

See the screenshot below that fixed my problem (hope it will help you).
![dissapearing-issue](https://user-images.githubusercontent.com/2078018/92326918-05009780-f056-11ea-9a51-d3264a697937.png)

**Comment by pranithasri** [comment]
why in some systems "+" sign is not available when connecting nodes.

**Comment by moklick** [comment]
> why in some systems "+" sign is not available when connecting nodes.
> 
> 

How is your question related to this issue?

**Comment by Harish-Hk** [comment]
@moklick Exactly, the stress works, but I don't know how to get my issue reproduced to a codepen or stackblitz.

Does it happen when the nodes overlap on one another?? but in the stress example, we do click change pos and that works fine. so not sure.

@100le  hey, thanks for your response, I did check my nodes positions, they're numbers...and nodes are rendering fine..its just the edges.

**Comment by moklick** [comment]
What are you doing so that this problem occurs? Are you moving nodes around? Are you adding nodes dynamically?

**Comment by Harish-Hk** [comment]
So I have a basic bootstrap tab approach, on opening the second tab, I'm mounting the react flow's parent component which render the nodes stored in the DB. switching to another tab unmounts the react-flow's parent component.

each rf node is basically another component.

![aug8](https://user-images.githubusercontent.com/28500085/92488101-bc8ada80-f20b-11ea-8cda-3a926cd8f7a0.JPG)

when I open second tab, the flow gets mounted and works fine, then I reload the whole page, which lands on first tab, and I again switch to second tab, the **first edge vanishes**. if I switch back to third or first tab again and come to second tab, it appears again.

Initially this happened with nodes, and as I already mentioned, the first node's opacity was being set to 0, which I set to 1 with the node's css classes.

I thought something similiar would crack it but, not sure why that happens.

**Comment by moklick** [comment]
Could you create a codesandox or repository with your tab/ReactFlow setup?

**Comment by moklick** [comment]
Hey @Harish-Hk!

any news about this issue? Otherwise I would like to close it.

**Comment by harish-madugula** [comment]
No @moklick , the issue is still there but I dont understand why it appears.
unfortunately my project is very big and I can't get a way to reproduce it in a shorter way.

**Comment by moklick** [comment]
Are you initializing the flow with elements stored in a database?

**Comment by moklick** [comment]
I would like to understand if it's a problem with ReactFlow or your implementation, but it's hard to tell when I can't see the code. If you want you could give me read permission for your repo or send me the code somehow.

**Comment by harish-madugula** [comment]
Hey, 
I observed that for some nodes, the `isInitialized`  is being set to false, due to which the edge from that node is disappered.

is there any relation to that property and the edge not being rendered?

I also see in the nodeModues index.ts files, that if super() is not called, the prop `isInitialized` is set to false.

![image](https://user-images.githubusercontent.com/28500085/94475470-bffef980-01ec-11eb-9704-01c3f43f1972.png)

is there any relation that react's super being deprecated causing this issue?

**Comment by moklick** [comment]
Are you wrapping each of your flows with the `ReactFlowProvider` component? If not, could you try that?

**Comment by harish-madugula** [comment]
Yup I tried that, no luck :(

**Comment by moklick** [comment]
We will close this issue for now. You can re-open it if you provide a code sandbox or something like that so that we can check the code. With no further information it's hard to tell what is going on here.

**Comment by mizkichan** [comment]
Hello, I have happened to encounter this issue while trying to figure out another problem.
https://codesandbox.io/s/vigorous-moser-i4be2
1. Click the 'add' button to create a new node.
2. Move the new node to somewhere else.
3. Click the 'add' button again and the existing node disappears because `opacity` is set to 0.

**Comment by moklick** [comment]
The ids need to be strings.

**Comment by aaronxdd** [comment]
I also face the issue, it not due to the opacity,when I change tab the flow render miss all the node and edge，You can see the pictures below for comparis.
![微信图片_20210924153509](https://user-images.githubusercontent.com/16145101/134636878-747296ef-5df6-4cc0-85f1-034c17e88378.png)
![微信图片_20210924153501](https://user-images.githubusercontent.com/16145101/134636913-f07eee3d-f92d-4256-81dc-26ec9fa0d114.png)
on

**Comment by aaronxdd** [comment]
can anyone help

**Comment by aaronxdd** [comment]
![微信图片_20210924154125](https://user-images.githubusercontent.com/16145101/134637174-e6e90672-a40c-477c-9e1c-2cd691db066f.png)

**Comment by aaronxdd** [comment]
I have solved the problem, add onlyRenderVisibleNodes={false} to your <ReactFlow/
![微信图片_20210930113551](https://user-images.githubusercontent.com/16145101/135383141-6ab2ab2f-40fd-4edb-a397-cce84e2ff59f.png)
>

**Comment by weisjohn** [comment]
> The ids need to be strings.

Had the same issue here too. It might be worth a console w/ a warning.

I know the docs say that they must be strings, however, the examples show `'1'`, `'2'`, etc.

**Comment by dominicdev** [comment]
I still have issue with this, any fix for this? onlyRenderVisibleNodes did not worked for me

**Comment by bcakmakoglu** [comment]
> I still have issue with this, any fix for this? onlyRenderVisibleNodes did not worked for me

Are you sure you're actually using strings as ids for your nodes? Numbers are not valid ids.

**Comment by ChenXiaoTemp** [comment]
I found the problem is useLayoutEffect , I delete it then it works

**Comment by karthiknmenon** [comment]
Not sure, but I think this is along the same lines.

I recently encountered an issue where edges were not rendering on the canvas, even though nodes were being populated correctly. The nodes and edges relied on data fetched from an API call, which was transformed on the frontend using various functions to ensure the desired structure.

At a certain point, I noticed that the edges would randomly disappear or reappear whenever the API was refetched using `tanstack-query`. After investigating, I discovered that the problem stemmed from the `setEdges` method being executed before the node positions and dimensions were calculated and set on the canvas using a layout algorithm, as described [here](https://reactflow.dev/learn/layouting/layouting).

I resolved this issue by ensuring that `setEdges` was only called after the API call was completed and the nodes were properly positioned on the canvas using the layout algorithm.

Hack - I also found that using `setTimeout` with a delay of `0` can resolve this issue effectively. This approach defers the execution of `setEdges`, allowing the browser to complete the layout and rendering of nodes before updating the edges. This small delay ensures that `setEdges` runs in the next event loop cycle, after the nodes are fully set up on the canvas.
